lib/timeout.rb

```diff
@@ -44,12 +44,107 @@ def self.handle_timeout(message) # :nodoc:
   end
 
   # :stopdoc:
-  CONDVAR = ConditionVariable.new
-  QUEUE = Queue.new
-  QUEUE_MUTEX = Mutex.new
-  TIMEOUT_THREAD_MUTEX = Mutex.new
-  @timeout_thread = nil
-  private_constant :CONDVAR, :QUEUE, :QUEUE_MUTEX, :TIMEOUT_THREAD_MUTEX
+
+  # We keep a private reference so that time mocking libraries won't break
+  # Timeout.
+  GET_TIME =
+    if defined?(Ractor.make_shareable)
+      begin
+        Ractor.make_shareable(Process.method(:clock_gettime))
+      rescue # failed on Ruby 3.4
+        Process.method(:clock_gettime)
+      end
+    else
+      Process.method(:clock_gettime)
+    end
+
+  private_constant :GET_TIME
+
+  class State
+    attr_reader :condvar, :queue, :queue_mutex # shared with Timeout.timeout()
+
+    def initialize
+      @condvar = ConditionVariable.new
+      @queue = Queue.new
+      @queue_mutex = Mutex.new
+
+      @timeout_thread = nil
+      @timeout_thread_mutex = Mutex.new
+    end
+
+    if defined?(Ractor.store_if_absent) &&
+       defined?(Ractor.shareble?) && Ractor.shareable?(GET_TIME)
+
+       # Ractor support if
+       # 1. Ractor.store_if_absent is available
+       # 2. Method object can be shareable (4.0~)
+
+       Ractor.store_if_absent :timeout_gem_state do
+         State.new
+       end
+
+       def self.instance
+         Ractor[:timeout_gem_state]
+       end
+
+       ::Timeout::RACTOR_SUPPORT = true # for test
+    else
+      @GLOBAL_STATE = State.new
```

Comment:
```suggestion
      GLOBAL_STATE = State.new
```
A constant should be faster (at least with JIT)